mlos_core/mlos_core/optimizers/observations.py

```diff
@@ -0,0 +1,361 @@
+        return len(self.config.index)
+
+    def __iter__(self) -> Iterator[Observations]:
+        for idx in self.config.index:
+            yield Observations(
+                config=self.config.loc[[idx]],
+                score=self.score.loc[[idx]],
+                context=None if self.context is None else self.context.loc[[idx]],
+                metadata=None if self.metadata is None else self.metadata.loc[[idx]],
+            )
+
+    def __repr__(self) -> str:
+        return (
+            f"Observation(config={self.config}, score={self.score},"
+            + " context={self.context}, metadata={self.metadata})"
+        )
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Observations):
+            return False
+
+        if not self.config.equals(other.config):
+            return False
+        if not self.score.equals(other.score):
+            return False
+        if not compare_optional_dataframe(self.context, other.context):
+            return False
+        if not compare_optional_dataframe(self.metadata, other.metadata):
+            return False
+
+        return True
+
+    def __ne__(self, other: Any) -> bool:
+        return not self.__eq__(other)
+
+
+class Suggestion:
+    """
+    A single suggestion for a configuration.
+
+    Attributes
+    ----------
+    config : pd.DataFrame
+        The suggested configuration.
+    """
+
+    def __init__(
+        self,
+        *,
+        config: pd.Series,
+        context: Optional[pd.Series] = None,
+        metadata: Optional[pd.Series] = None,
+    ):
+        self.config = config
+        self.context = context
+        self.metadata = metadata
+
+    def complete(self, score: pd.Series) -> Observation:
+        """
+        Completes the suggestion.
+
+        Parameters
+        ----------
+        score : pd.Series
+            The score metrics observed.
+
+        Returns
+        -------
+        Observation
+            The observation of the suggestion.
+        """
+        return Observation(
+            config=self.config,
+            score=score,
+            context=self.context,
+            metadata=self.metadata,
+        )
+
+    def config_to_configspace(self, space: ConfigurationSpace) -> Configuration:
+        return Configuration(space, values=self.config.to_dict())
+
+    def __repr__(self) -> str:
+        return (
+            f"Suggestion(config={self.config}, context={self.context}, metadata={self.metadata})"
+        )
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Suggestion):
+            return False
+
+        if not self.config.equals(other.config):
+            return False
+        if not compare_optional_series(self.context, other.context):
+            return False
+        if not compare_optional_series(self.metadata, other.metadata):
+            return False
+
+        return True
+
+    def __ne__(self, other: Any) -> bool:
+        return not self.__eq__(other)
```

Comment:
Is this necessary?  Or is it automatically handled?

Reply:
This is required for something I did, I can't remember exactly where, but it is not automatic.